webapp/channels/src/utils/dialog_conversion.ts

```diff
@@ -449,13 +497,147 @@ export function convertDialogToAppForm(
         submit: {
             path: '/submit',
             expand: {},
+            state: dialogState || undefined,
         },
         fields: convertedFields,
     };
 
+    // Set source if sourceUrl is provided or if any fields have refresh enabled
+    const hasRefreshFields = convertedFields.some((field) => field.refresh === true);
+    if ((sourceUrl && sourceUrl.trim()) || hasRefreshFields) {
+        form.source = {
+            path: sourceUrl || '/refresh',
+            expand: {},
+            state: dialogState || undefined,
+        };
+    }
+
     return {form, errors: allErrors};
 }
 
+/**
+ * Extract primitive values from form field objects for storage/submission
+ * Converts select option objects {label: "Text", value: "val"} to primitive "val"
+ * Filters out null, undefined, empty, and "<nil>" values
+ */
+export function extractPrimitiveValues(values: Record<string, any>): Record<string, any> {
```

Comment:
Not sure if it's necessary but have we considered memoizing any of these functions? They seem expensive